### PR TITLE
Add memory location comparison macros (ck_assert_mem_*)

### PR DESCRIPTION
**Pull request by prusnak**
As discussed here: https://sourceforge.net/p/check/feature-requests/30/

### CODE REVIEW

**Comment by brarcher** [comment]
Jenkins: ok to test
^ necessary to start the remaining of the tests

File: src/check.h.in [hunk 1]
```diff
@@ -747,6 +747,108 @@ CK_DLL_EXP void CK_EXPORT _ck_assert_failed(const char *file, int line,
  */
 #define ck_assert_str_ge(X, Y) _ck_assert_str(X, >=, Y)
 
+/* Memory location comparison macros with improved output compared to ck_assert() */
+/* OP might be any operator that can be used in '0 OP memcmp(X,Y,L)' comparison */
+/* The x and y parameter swap in memcmp() is needed to handle >, >=, <, <= operators */
+/* Output is limited to 64 bytes (or 128 hex digits) */
+#define _ck_assert_mem(X, OP, Y, L) do { \
+  const char* _ck_x = (const char*)(void*)(X); \
```
Comment (brarcher) [inline src/check.h.in]: why is the (void*) cast necessary?

File: src/check.h.in [hunk 2]
```diff
@@ -747,6 +747,108 @@ CK_DLL_EXP void CK_EXPORT _ck_assert_failed(const char *file, int line,
...
+  for (_ck_i = 0; _ck_i < ((_ck_l > 64) ? 64 : _ck_l); _ck_i++) { \
+    _ck_x_str[_ck_i * 2  ]   = _ck_hexdigits[(_ck_x[_ck_i] >> 4) & 0xF]; \
+    _ck_y_str[_ck_i * 2  ]   = _ck_hexdigits[(_ck_y[_ck_i] >> 4) & 0xF]; \
+    _ck_x_str[_ck_i * 2 + 1] = _ck_hexdigits[_ck_x[_ck_i] & 0xF]; \
+    _ck_y_str[_ck_i * 2 + 1] = _ck_hexdigits[_ck_y[_ck_i] & 0xF]; \
+  } \
+  _ck_x_str[_ck_i * 2] = 0; \
+  _ck_y_str[_ck_i * 2] = 0; \
+  ck_assert_msg(0 OP memcmp(_ck_y, _ck_x, _ck_l), \
+    "Assertion '"#X#OP#Y"' failed: "#X"==\"%s\", "#Y"==\"%s\"", _ck_x_str, _ck_y_str); \
```
Comment (brarcher) [inline src/check.h.in]: I am torn between the message containing "#X#OP#Y" or something else which mentions the size of the strings. Should the message mention memcmp, as the documentation uses? For example, "0#OPmemcmp(#X,#Y,#L)"?

File: src/check.h.in [hunk 3]
```diff
@@ -747,6 +747,108 @@ CK_DLL_EXP void CK_EXPORT _ck_assert_failed(const char *file, int line,
...
 
+/* Memory location comparison macros with improved output compared to ck_assert() */
+/* OP might be any operator that can be used in '0 OP memcmp(X,Y,L)' comparison */
+/* The x and y parameter swap in memcmp() is needed to handle >, >=, <, <= operators */
+/* Output is limited to 64 bytes (or 128 hex digits) */
+#define _ck_assert_mem(X, OP, Y, L) do { \
+  const char* _ck_x = (const char*)(void*)(X); \
+  const char* _ck_y = (const char*)(void*)(Y); \
+  size_t _ck_l = (L); \
+  char _ck_x_str[129]; \
```
Comment (brarcher) [inline src/check.h.in]: Can the magic numbers used in the macro, e.g. "129", "64", be defined? For example CK_MAX_ASSERT_MEM_PRINT_SIZE=64 which is then used everywhere?

File: src/check.h.in (hunk 1 above)
Comment (brarcher) [inline src/check.h.in]: Just to be explicit, in case there is a platform where char!=8bits, should this be uint8_t?

File: src/check.h.in [hunk 4]
```diff
@@ -747,6 +747,108 @@ CK_DLL_EXP void CK_EXPORT _ck_assert_failed(const char *file, int line,
...
+  char _ck_y_str[129]; \
+  static char _ck_hexdigits[] = "0123456789abcdef"; \
```
Comment (brarcher) [inline src/check.h.in]: Could this array be made const?

File: src/check.h.in [hunk 5]
```diff
@@ -747,6 +747,108 @@ CK_DLL_EXP void CK_EXPORT _ck_assert_failed(const char *file, int line,
...
+  size_t _ck_i; \
+  for (_ck_i = 0; _ck_i < ((_ck_l > 64) ? 64 : _ck_l); _ck_i++) { \
```
Comment (brarcher) [inline src/check.h.in]: can the computation for the string size be done before the loop, for readability?

File: src/check.h.in (hunk 3 above)
Comment (brarcher) [inline src/check.h.in]: Additionally, perhaps define it as such:

```
#ifndef CK_MAX_ASSERT_MEM_PRINT_SIZE
#define CK_MAX_ASSERT_MEM_PRINT_SIZE 
#endif
```

That way if a user wishes to override it they may. Another option it to make it a configure option

File: src/check.h.in (hunk 1 above)
Comment (prusnak) [inline src/check.h.in]: Changed to uint8_t and removed void* cast.

File: src/check.h.in (hunk 3 above)
Comment (prusnak) [inline src/check.h.in]: Good idea.

File: src/check.h.in (hunk 4 above)
Comment (prusnak) [inline src/check.h.in]: Yes!

File: src/check.h.in (hunk 5 above)
Comment (prusnak) [inline src/check.h.in]: OK

**Comment by prusnak** [comment]
Pushed a new commit fixing the issues from above.

**Comment by prusnak** [comment]
Also I am adding the following for discussion. The way I usually use these macros is like this:

```
ck_assert_mem_eq(der, fromhex("3007020166020200ee"), 9); 
```

while the function fromhex is defined like this:

```
#define FROMHEX_MAXLEN 256
const uint8_t *fromhex(const char *str)
{
	static uint8_t buf[FROMHEX_MAXLEN];
	size_t len = strlen(str) / 2;
	if (len > FROMHEX_MAXLEN) len = FROMHEX_MAXLEN;
	for (size_t i = 0; i < len; i++) {
		uint8_t c = 0;
		if (str[i*2] >= '0' && str[i*2] <= '9') c += (str[i*2] - '0') << 4;
		if (str[i*2] >= 'a' && str[i*2] <= 'f') c += (10 + str[i*2] - 'a') << 4;
		if (str[i*2] >= 'A' && str[i*2] <= 'F') c += (10 + str[i*2] - 'A') << 4;
		if (str[i*2+1] >= '0' && str[i*2+1] <= '9') c += (str[i*2+1] - '0');
		if (str[i*2+1] >= 'a' && str[i*2+1] <= 'f') c += (10 + str[i*2+1] - 'a');
		if (str[i*2+1] >= 'A' && str[i*2+1] <= 'F') c += (10 + str[i*2+1] - 'A');
		buf[i] = c;
	}
	return buf;
}
```

Do you think it makes sense to integrate it further in the check library, so we could use something like this?

```
ck_assert_mem_hex_eq(der, "3007020166020200ee", 9); 
```

It is much more convenient (especially for longer values, eg. 64 bytes) than using:

```
ck_assert_mem_hex_eq(der, "\x30\x07\x02\x01\x66\x02\x02\x00\xee", 9); 
```

**Comment by brarcher** [comment]
I can appreciate the use case you mention regarding the hex. What gives me pause is it would be another place where a maximum size would be used (for the static buffer). The first place using a max size, the strings which reports buffer contents on a failure, does not result in a problem if more data is used than the buffer size. In that case the differences in the two buffers will simply not be printed, but it will not affect the success or failure of the comparison. However, in the case of the hex based comparison call because it may not be obvious to a caller that the hex string provided has a maximum possible length, passing too many characters may affect the expected result. In addition, if an odd number of characters were passed the last character would be ignored.

From projects where I've compared binary data typically the expected data is in a file, and before the comparison the file's contents are loaded into a buffer. However, those typically are for comparisons that are larger than one would want to type for a unit test.

My opinion is to provide the ck_assert_mem_* calls for now, and if there is a way to quell the concerns listed above maybe a ck_assert_hex call can be added as a separate commit.

**Comment by prusnak** [comment]
Ok, let's focus on `ck_assert_mem_*` for now. Do you want some other changes or the commit is good as it is now?

**Comment by brarcher** [comment]
> Do you want some other changes or the commit is good as it is now?

The change will need to have some unit tests accompany it. You can add some test cases under the test_ck_assert_ptr_ne test here:

https://github.com/libcheck/check/blob/master/tests/check_check_sub.c#L555

You will also need to add the test cases around here:

https://github.com/libcheck/check/blob/master/tests/check_check_sub.c#L1098

After that, you will need to determine the expected result of each of the test cases (there should be some tests that pass the check and some that fail) and put the expected result at the end of the pointer-based tests in the master_tests table:

https://github.com/libcheck/check/blob/master/tests/check_check_master.c#L117

Just keep the order of adding the tests to the _simple_ test case and the order in the master_tests table consistent to avoid issues. Look around check_check_sub.c for examples of other tests for the ck_assert* macros for inspiration.

After adding those tests the change should be good to go. Let me know if you have any questions or run into something when adding the unit tests.

**Comment by prusnak** [comment]
Added unittests as requested in f9676cf ...

File: tests/check_check_sub.c [hunk 6]
```diff
@@ -553,6 +553,78 @@ START_TEST(test_ck_assert_ptr_ne)
 }
 END_TEST
 
+START_TEST(test_ck_assert_mem_eq)
```
Comment (brarcher) [inline tests/check_check_sub.c]: Because there is a limit on the number of characters printed in the assert message, there will need to be a test that check the boundary conditions of the message. Namely, a test comparing 0 bytes (should never produce a message), a test where all the bytes are printed (covered already), a test which is the max number of bytes allowed, and a test with too many bytes. Can you add tests for these cases?

**Comment by brarcher** [comment]
It would appear that one of the unit tests is not passing. [This example](https://travis-ci.org/libcheck/check/jobs/170772788) shows the following message in the test run output:
```
/home/travis/build/libcheck/check/tests/check_check_master.c:473:F:Core Tests:test_check_all_msgs:56: Expected Assertion '", got Assertion '"\x00\x00\x00\x00\x01" == s' failed: "\x00\x00\x00\x00\x01" == "0000000001", s == "0000000002"
```
I believe that means there was no failure expected but a failure did occur. I am not sure the reason for this.

If you would like to reproduce this locally, there is a way to disable the unit tests which take a long time to run. Give the --disable-timeout-tests argument to configure and then the unit tests should run fairly quickly. You could also execute the check_check unit test program directly and check for a non-zero return code.

**Comment by prusnak** [comment]
> I believe that means there was no failure expected but a failure did occur. I am not sure the reason for this.

Ah, there was an error in `check_check_master.c`. Should have used double backslash in C string.

Also added two new tests: `test_ck_assert_mem_zerolen` and `test_ck_assert_mem_eq_long`.

File: tests/check_check_sub.c [hunk 7]
```diff
@@ -553,6 +553,95 @@ START_TEST(test_ck_assert_ptr_ne)
...
+  const char *s = "\x00\x00\x00\x00\x01";
+  const char *t = "\x00\x00\x00\x00\x02";
+  ck_assert_mem_ge(t, s, 5);
+  ck_assert_mem_ge(t, t, 5);
+  record_failure_line_num(__LINE__);
+  ck_assert_mem_ge(s, t, 5);
+}
+END_TEST
+
+START_TEST(test_ck_assert_mem_expr)
```
Comment (brarcher) [inline tests/check_check_sub.c]: what does this test for which the other test do not already test for?

File: tests/check_check_sub.c [hunk 8]
```diff
@@ -553,6 +553,95 @@ START_TEST(test_ck_assert_ptr_ne)
...
+{
+  const char *s = "\x00\x00\x00\x00\x01";
+  const char *t[] = { "\x00\x00\x00\x00\x01", "\x00\x00\x00\x00\x02" };
+  int i = -1;
+  ck_assert_str_eq(s, t[++i]);
+  ck_assert_str_eq(s, t[i]);
```
Comment (brarcher) [inline tests/check_check_sub.c]: Are not both of these identical? Is not t[1] never used?

File: tests/check_check_sub.c [hunk 9]
```diff
@@ -553,6 +553,95 @@ START_TEST(test_ck_assert_ptr_ne)
...
+  const char *s = "\x00\x00\x00\x00\x02";
+  const char *t = "\x00\x00\x00\x00\x01";
+  ck_assert_mem_eq(t, s, 0);
+}
+END_TEST
+
+START_TEST(test_ck_assert_mem_eq_long)
+{
+  const char *s = "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x02";
+  const char *t = "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01";
```
Comment (brarcher) [inline tests/check_check_sub.c]: To easily check that the two arrays are both at least 65 characters long, might it be easier to:

const char s[65] = {0};
const char t[65] = {1};

File: tests/check_check_master.c [hunk 10]
```diff
@@ -114,6 +114,15 @@ static master_test_t master_tests[] = {
...
+  { "Simple Tests", CK_FAILURE, "Assertion '\"\\x00\\x00\\x00\\x00\\x01\" == s' failed: \"\\x00\\x00\\x00\\x00\\x01\" == \"0000000001\", s == \"0000000002\"" },
+  { "Simple Tests", CK_FAILURE, "Assertion 't != s' failed: t == \"0000000002\", s == \"0000000002\"" },
+  { "Simple Tests", CK_FAILURE, "Assertion 's < s' failed: s == \"0000000001\", s == \"0000000001\"" },
+  { "Simple Tests", CK_FAILURE, "Assertion 't <= s' failed: t == \"0000000002\", s == \"0000000001\"" },
+  { "Simple Tests", CK_FAILURE, "Assertion 't > t' failed: t == \"0000000002\", t == \"0000000002\"" },
+  { "Simple Tests", CK_FAILURE, "Assertion 's >= t' failed: s == \"0000000001\", t == \"0000000002\"" },
+  { "Simple Tests", CK_PASS,    "Passed" },
+  { "Simple Tests", CK_PASS,    "Passed" },
+  { "Simple Tests", CK_FAILURE, "Assertion 't == s' failed: t == \"00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000\", s == \"00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000\"" },
```
Comment (brarcher) [inline tests/check_check_master.c]: Here is a question about the output of an assertion when the data is too long. In this example we see that both the first and second argument look the same though the data is different. When the data is too long and it is truncated should a "..." be added to the end of the two arguments, so it is obvious to the developer that the difference may not be in the portion of the string shown?

File: tests/check_check_sub.c (hunk 9 above)
Comment (prusnak) [inline tests/check_check_sub.c]: Yes, but I need the different character at the end of the array, not at the beginning.

File: tests/check_check_sub.c (hunk 7 above)
Comment (prusnak) [inline tests/check_check_sub.c]: Removed.

File: tests/check_check_master.c (hunk 10 above)
Comment (prusnak) [inline tests/check_check_master.c]: Added in 8b1cd61

File: tests/check_check_sub.c (hunk 9 above)
Comment (brarcher) [inline tests/check_check_sub.c]: That is fine, it is OK as is.

File: tests/check_check_master.c [hunk 11]
```diff
@@ -121,7 +121,7 @@ static master_test_t master_tests[] = {
   { "Simple Tests", CK_FAILURE, "Assertion 't > t' failed: t == \"0000000002\", t == \"0000000002\"" },
   { "Simple Tests", CK_FAILURE, "Assertion 's >= t' failed: s == \"0000000001\", t == \"0000000002\"" },
   { "Simple Tests", CK_PASS,    "Passed" },
-  { "Simple Tests", CK_FAILURE, "Assertion 't == s' failed: t == \"00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000\", s == \"00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000\"" },
+  { "Simple Tests", CK_FAILURE, "Assertion 't == s' failed: t == \"000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000..\", s == \"000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000..\"" },
```
Comment (brarcher) [inline tests/check_check_master.c]: As there would be a difference between one check against CK_MAX_ASSERT_MEM_PRINT_SIZE characters and one with CK_MAX_ASSERT_MEM_PRINT_SIZE+1 characters, can there be a test using exactly CK_MAX_ASSERT_MEM_PRINT_SIZE characters to show that there is no .. at the end?

File: src/check.h.in [hunk 12]
```diff
@@ -772,6 +772,12 @@ CK_DLL_EXP void CK_EXPORT _ck_assert_failed(const char *file, int line,
   } \
   _ck_x_str[_ck_i * 2] = 0; \
   _ck_y_str[_ck_i * 2] = 0; \
+  if (_ck_i != _ck_l) { \
```
Comment (prusnak) [inline src/check.h.in]: OK.

File: tests/check_check_master.c (hunk 11 above)
Comment (prusnak) [inline tests/check_check_master.c]: Added.

**Comment by brarcher** [comment]
I think that this change is good to go, thanks for the contribution.

Would you mind adding documentation for this change [here](https://github.com/libcheck/check/blob/master/doc/check.texi#L989), perhaps in a separate commit? Maybe add it after the description of ck_assert_ptr_eq and ck_assert_ptr_ne.

I'll look into the unit test which is failing. It appears to be failing before this change, so I believe it to be unrelated.

**Comment by brarcher** [comment]
I think that this change is good to go, thanks for the contribution.

Would you mind adding documentation for this change [here](https://github.com/libcheck/check/blob/master/doc/check.texi#L989), perhaps in a separate commit? Maybe add it after the description of ck_assert_ptr_eq and ck_assert_ptr_ne.

I'll look into the unit test which is failing. It appears to be failing before this change, so I believe it to be unrelated.

**Comment by prusnak** [comment]
Added documentation in a separate commit.

**Comment by brarcher** [comment]
Thanks for the contribution!

**Comment by brarcher** [comment]
Oh, I forgot to ask, @prusnak . If you are interested would you mind submitting a change for the [AUTHORS](https://github.com/libcheck/check/blob/master/AUTHORS) file, adding an entry for yourself?

**Comment by prusnak** [comment]
Added in https://github.com/libcheck/check/pull/68
